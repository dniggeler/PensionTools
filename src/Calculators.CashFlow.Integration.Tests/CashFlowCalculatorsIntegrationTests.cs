using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Mvc.Testing;
using Microsoft.AspNetCore.TestHost;
using Microsoft.Extensions.Configuration;
using PensionCoach.Tools.CommonTypes;
using PensionCoach.Tools.TaxComparison;
using Snapshooter.Xunit;
using TaxCalculator.WebApi;
using Xunit;

namespace Calculators.CashFlow.Integration.Tests
{
    [Trait("Cash-Flow", "Integration")]
    public class CashFlowCalculatorsIntegrationTests : IClassFixture<WebApplicationFactory<Startup>>
    {
        private readonly HttpClient client;

        public CashFlowCalculatorsIntegrationTests()
        {
            var testServer = new TestServer(
                new WebHostBuilder()
                    .ConfigureAppConfiguration((_, builder) =>
                    {
                        builder.AddJsonFile("appsettings.integration.json");
                    })
                    .UseStartup<Startup>());

            client = testServer.CreateClient();
            client.BaseAddress = new Uri("http://localhost/api/scenario/tax/");
        }

        [Fact(DisplayName = "Default")]
        public async Task ShouldCalculateSuccessfully()
        {
            CapitalBenefitTransferInComparerRequest request = GetRequest();

            var response = await client.PostAsJsonAsync("CalculateTransferInCapitalBenefits", request);

            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadAsStringAsync();

            Snapshot.Match(result);
        }

        private static CapitalBenefitTransferInComparerRequest GetRequest()
        {
            int yearOfCapitalBenefitWithdrawal = 2030;

            return new CapitalBenefitTransferInComparerRequest
            {
                Name = "Test Multi-Period Calculator",
                CalculationYear = 2021,
                BfsMunicipalityId = 261,
                CivilStatus = CivilStatus.Married,
                ReligiousGroup = ReligiousGroupType.Other,
                PartnerReligiousGroup = ReligiousGroupType.Other,
                TaxableIncome = 100_000,
                TaxableFederalIncome = 100_000,
                TaxableWealth = 500_000,
                TransferIns = new List<SingleTransferInModel>
                {
                    new(10_000, new DateTime(2021, 1, 1))
                },
                WithCapitalBenefitTaxation = true,
                Withdrawals = new List<SingleTransferInModel>
                {
                    new(500_000, new DateTime(yearOfCapitalBenefitWithdrawal, 1, 1))
                },
            };
        }
    }
}
