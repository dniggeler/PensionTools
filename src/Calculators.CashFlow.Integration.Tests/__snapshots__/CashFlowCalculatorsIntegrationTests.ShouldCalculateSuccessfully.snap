{"startingYear":2021,"numberOfPeriods":10,"accounts":[{"year":2021,"amount":516778.865,"accountType":"Wealth"},{"year":2021,"amount":0,"accountType":"CapitalBenefits"},{"year":2022,"amount":533597.49000,"accountType":"Wealth"},{"year":2022,"amount":0,"accountType":"CapitalBenefits"},{"year":2023,"amount":550456.6850000,"accountType":"Wealth"},{"year":2023,"amount":0,"accountType":"CapitalBenefits"},{"year":2024,"amount":567359.480000000,"accountType":"Wealth"},{"year":2024,"amount":0,"accountType":"CapitalBenefits"},{"year":2025,"amount":584302.93530000000,"accountType":"Wealth"},{"year":2025,"amount":0,"accountType":"CapitalBenefits"},{"year":2026,"amount":601262.6218030000000,"accountType":"Wealth"},{"year":2026,"amount":0,"accountType":"CapitalBenefits"},{"year":2027,"amount":618265.181321030000000,"accountType":"Wealth"},{"year":2027,"amount":0,"accountType":"CapitalBenefits"},{"year":2028,"amount":635290.24688424030000000,"accountType":"Wealth"},{"year":2028,"amount":0,"accountType":"CapitalBenefits"},{"year":2029,"amount":652364.5230530827030000000,"accountType":"Wealth"},{"year":2029,"amount":0,"accountType":"CapitalBenefits"},{"year":2030,"amount":669467.706233613530030000000,"accountType":"Wealth"},{"year":2030,"amount":0,"accountType":"CapitalBenefits"}]}
